File: structural/composite/python/composite.py
from typing import List
from component import FileSystemComponent


class Directory(FileSystemComponent):
    def __init__(self, name: str):
        self.name = name
        self.children: List[FileSystemComponent] = []

    def add(self, component: FileSystemComponent) -> None:
        self.children.append(component)

    def remove(self, component: FileSystemComponent) -> None:
        self.children = [child for child in self.children if child != component]

    def get_name(self) -> str:
        return self.name

    def get_size(self) -> int:
        return sum(child.get_size() for child in self.children)

    def print(self) -> None:
        print(f"Directory: {self.name} [Size: {self.get_size()}]")
        for child in self.children:
            if isinstance(child, Directory):
                child.print()
            else:
                print(f"  File: {child.get_name()} [Size: {child.get_size()}]")

File: structural/composite/typescript/composite.ts
import { FileSystemComponent } from "./component";

export class Directory implements FileSystemComponent {
    private children: FileSystemComponent[] = [];

    constructor(private name: string) {}

    add(component: FileSystemComponent): void {
        this.children.push(component);
    }

    remove(component: FileSystemComponent): void {
        this.children = this.children.filter(child => child !== component);
    }

    getName(): string {
        return this.name;
    }

    getSize(): number {
        return this.children.reduce((acc, child) => acc + child.getSize(), 0);
    }

    print(): void {
        console.log(`Directory: ${this.name} [Size: ${this.getSize()}]`);
        this.children.forEach(child => {
            if (child instanceof Directory) {
                child.print();
            } else {
                console.log(`  File: ${child.getName()} [Size: ${child.getSize()}]`);
            }
        });
    }
}